vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt

```diff
@@ -101,14 +122,65 @@ class VesselImpl(
      */
     private var closeWasCalled: Boolean = false
 
+    private val profiler: Profiler = when {
+        profile -> ProfilerImpl()
+        else -> DummyProfiler()
+    }
+
+    private fun preloadImpl(entities: List<VesselEntity>) {
+        for (entity in entities) {
+            val kclass = try {
+                Class.forName(entity.type).kotlin
+            } catch (error: Exception) {
+                continue
+            }
+
+            val data = if (entity.data != null) fromJson(entity.data, kclass) else nullValue
+            cache?.set(entity.type, data as Any)
+        }
+    }
+
+    override suspend fun preload() {
+        if (cache == null) {
+            return
+        }
+
+        val entities = profiler.time(Span.PRELOAD_FROM_DB) {
+            dao.getAll()
+        }
+
+        preloadImpl(entities)
+    }
+
+    override fun preloadBlocking() {
+        if (cache == null) {
+            return
+        }
+
+        val entities = profiler.timeBlocking(Span.PRELOAD_FROM_DB) {
+            dao.getAllBlocking()
+        }
+
+        preloadImpl(entities)
+    }
+
+    // endregion
+
+    // region profiling
+
+    override val profileData
+        get() = profiler.snapshot
+
     // endregion
 
     // region helper functions
 
     /**
      * Convert a stored json string into the specified data type.
      */
-    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? = gson.fromJson<T>(value, type.java)
+    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? {
```

Comment:
this change should be undone. idiomatic kotlin to use = instead of block with single line return